src/core/thread/anycast_locator.cpp

```diff
@@ -0,0 +1,163 @@
+ */
+
+#include "anycast_locator.hpp"
+
+#if OPENTHREAD_CONFIG_TMF_ANYCAST_LOCATOR_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/instance.hpp"
+#include "common/locator_getters.hpp"
+#include "thread/thread_tlvs.hpp"
+#include "thread/uri_paths.hpp"
+
+namespace ot {
+
+AnycastLocator::AnycastLocator(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mCallback(nullptr)
+    , mContext(nullptr)
+    , mAnycastLocate(UriPath::kAnycastLocate, HandleAnycastLocate, this)
+
+{
+    Get<Tmf::Agent>().AddResource(mAnycastLocate);
+}
+
+Error AnycastLocator::Locate(const Ip6::Address &aAnycastAddress, Callback aCallback, void *aContext)
+{
+    Error            error   = kErrorNone;
+    Coap::Message *  message = nullptr;
+    Ip6::MessageInfo messageInfo;
+
+    VerifyOrExit((aCallback != nullptr) && Get<Mle::Mle>().IsAnycastLocator(aAnycastAddress),
+                 error = kErrorInvalidArgs);
+
+    message = Get<Tmf::Agent>().NewMessage();
+    VerifyOrExit(message != nullptr, error = kErrorNoBufs);
+
+    SuccessOrExit(error = message->InitAsConfirmablePost(UriPath::kAnycastLocate));
+    SuccessOrExit(error = message->SetPayloadMarker());
+
+    if (mCallback != nullptr)
+    {
+        IgnoreError(Get<Tmf::Agent>().AbortTransaction(HandleResponse, this));
+    }
+
+    messageInfo.SetSockAddr(Get<Mle::MleRouter>().GetMeshLocal16());
+    messageInfo.SetPeerAddr(aAnycastAddress);
+    messageInfo.SetPeerPort(Tmf::kUdpPort);
+
+    SuccessOrExit(error = Get<Tmf::Agent>().SendMessage(*message, messageInfo, HandleResponse, this));
+
+    mCallback = aCallback;
+    mContext  = aContext;
+
+exit:
+    FreeMessageOnError(message, error);
+    return error;
+}
+
+void AnycastLocator::HandleResponse(void *               aContext,
+                                    otMessage *          aMessage,
+                                    const otMessageInfo *aMessageInfo,
+                                    Error                aError)
+{
+    static_cast<AnycastLocator *>(aContext)->HandleResponse(
+        static_cast<Coap::Message *>(aMessage), static_cast<const Ip6::MessageInfo *>(aMessageInfo), aError);
+}
+
+void AnycastLocator::HandleResponse(Coap::Message *aMessage, const Ip6::MessageInfo *aMessageInfo, Error aError)
+{
+    OT_UNUSED_VARIABLE(aMessageInfo);
+
+    uint16_t            rloc16  = Mac::kShortAddrInvalid;
+    const Ip6::Address *address = nullptr;
+    Ip6::Address        meshLocalAddress;
+
+    SuccessOrExit(aError);
+    OT_ASSERT(aMessage != nullptr);
+
+    meshLocalAddress.SetPrefix(Get<Mle::Mle>().GetMeshLocalPrefix());
+    SuccessOrExit(Tlv::Find<ThreadMeshLocalEidTlv>(*aMessage, meshLocalAddress.GetIid()));
+    SuccessOrExit(Tlv::Find<ThreadRloc16Tlv>(*aMessage, rloc16));
+
+#if OPENTHREAD_FTD
+    if (Get<AddressResolver>().UpdateCacheEntry(meshLocalAddress, rloc16) != kErrorNone)
+    {
+        Get<AddressResolver>().AddSnoopedCacheEntry(meshLocalAddress, rloc16);
+    }
```

Comment:
We used more complex conditions for anoter calling to `AddSnoopedCacheEntry`:
```
        if (Get<AddressResolver>().UpdateCacheEntry(ip6Header.GetSource(), aMeshSource.GetShort()) == kErrorNotFound)
        {
            // Thread 1.1 Specification 5.5.2.2: FTDs MAY add/update
            // EID-to-RLOC Map Cache entries by inspecting packets
            // being received. We exclude frames from an MTD child
            // source and verify that the destination is the device
            // itself or an MTD child of the device.

            if (Get<Mle::MleRouter>().IsFullThreadDevice() &&
                !Get<Mle::MleRouter>().IsMinimalChild(aMeshSource.GetShort()) &&
                (aMeshDest.GetShort() == Get<Mac::Mac>().GetShortAddress() ||
                 Get<Mle::MleRouter>().IsMinimalChild(aMeshDest.GetShort())))
            {
                Get<AddressResolver>().AddSnoopedCacheEntry(ip6Header.GetSource(), aMeshSource.GetShort());
            }
        }
```

Is it necessary to check for `Get<Mle::MleRouter>().IsFullThreadDevice() && !Get<Mle::MleRouter>().IsMinimalChild(aMeshSource.GetShort())` here?

Reply:
I noticed the same earlier. I actually wanted to simplify and move this common check into `AddressResolver` itself.  Just submitted PR https://github.com/openthread/openthread/pull/6604.

Reply:
> I noticed the same earlier. I actually wanted to simplify and move this common check into `AddressResolver` itself. Just submitted PR #6604.

Same thought.